src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs

```diff
@@ -143,7 +143,23 @@ internal static void ExecuteBackoutCodeHelper(object backoutCode, object userDat
             ((CleanupCode)backoutCode)(userData, exceptionThrown);
         }
 
-        /// <returns>true if given type is reference type or value type that contains references</returns>
+        /// <summary>
+        /// Returns true if the given type is a reference type (e.g., class, interface);
+        /// false if the given type is a value type (e.g., int, enum, ArraySegment).
+        /// </summary>
+        internal static bool IsReference<T>()
```

Comment:
This could be useful on its own.

Reply:
Yeah, I'd like to plumb this through `Span<T>` and `SpanHelpers` as a best practice. Consider the line below (from the implementation of `MemoryExtensions.Contains`).

https://github.com/dotnet/coreclr/blob/311b5e2fe413c6c74a2a3680ab54d8a978651472/src/System.Private.CoreLib/shared/System/SpanHelpers.T.cs#L60

This check makes me nervous because `default(T) == null` passes for `Nullable<T>`, and the only thing that saves this code from going down the wrong branch and performing out-of-bounds reads is the fact that `Nullable<T>` can't implement `IEquatable<T>`, so there's a compile-time error. If this code gets copied and pasted to a method that doesn't have that generic constraint then type safety could be violated.

Reply:
Then may be it is time for  https://github.com/dotnet/coreclr/issues/2591?